src/main/kotlin/app/revanced/patcher/apk/Apk.kt

```diff
@@ -0,0 +1,353 @@
+@file:Suppress("MemberVisibilityCanBePrivate")
+
+package app.revanced.patcher.apk
+
+import app.revanced.patcher.DomFileEditor
+import app.revanced.patcher.Patcher
+import app.revanced.patcher.PatcherOptions
+import app.revanced.patcher.arsc.*
+import app.revanced.patcher.util.ProxyBackedClassList
+import app.revanced.patcher.util.xml.LazyXMLInputSource
+import app.revanced.patcher.util.xml.scanIdRegistrations
+import com.reandroid.apk.ApkModule
+import com.reandroid.apk.xmlencoder.EncodeException
+import com.reandroid.apk.xmlencoder.EncodeMaterials
+import com.reandroid.apk.xmlencoder.EncodeUtil
+import com.reandroid.archive.InputSource
+import com.reandroid.arsc.chunk.PackageBlock
+import com.reandroid.arsc.chunk.TableBlock
+import com.reandroid.arsc.chunk.xml.AndroidManifestBlock
+import com.reandroid.arsc.value.Entry
+import com.reandroid.arsc.value.ResConfig
+import lanchon.multidexlib2.BasicDexEntry
+import lanchon.multidexlib2.DexIO
+import lanchon.multidexlib2.MultiDexContainerBackedDexFile
+import lanchon.multidexlib2.MultiDexIO
+import lanchon.multidexlib2.RawDexIO
+import org.jf.dexlib2.dexbacked.DexBackedDexFile
+import org.jf.dexlib2.iface.MultiDexContainer
+import org.jf.dexlib2.writer.io.MemoryDataStore
+import org.w3c.dom.*
+import java.io.File
+import java.util.zip.ZipEntry
+
+
+/**
+ * The [Apk] file that is to be patched.
+ *
+ * @param path The path to the apk file.
+ * @param name The name of this apk.
+ */
+sealed class Apk private constructor(internal val module: ApkModule) {
+    companion object {
+        const val manifest = "AndroidManifest.xml"
+    }
+
+    /**
+     * The metadata of the [Apk].
+     */
+    val packageMetadata = PackageMetadata(module.androidManifestBlock)
+
+    internal val archive = Archive(module)
+
+
+    /**
+     * Refresh updated resources for an [Apk].
+     *
+     * @param options The [PatcherOptions] of the [Patcher].
+     */
+    internal open fun finalize(options: PatcherOptions) {
+        archive.openFiles.forEach { options.logger.warn("File $it was never closed! File modifications will not be applied if you do not close them.") }
+
+        resources.useMaterials {
+            module.apkArchive.listInputSources().forEach {
+                val pkg = resources.packageBlock
+                if (it is LazyXMLInputSource) {
+                    if (resources.hasResourceTable) {
+                        // Scan for @+id registrations.
+                        it.document.scanIdRegistrations().forEach { attr ->
+                            val name = attr.value.split('/').last()
+                            options.logger.trace("Registering ID: $name")
+                            pkg!!.getOrCreate("", "id", name).resValue.valueAsBoolean = false
```

Comment:
Please document this codeblock from the perspective of the patcher. It is unknown what is going on here or why it is necessary.

Reply:
Fair. I think I can refactor it to be more clear.